### PR TITLE
Add confirmation popup on exit and return to menu

### LINKED ISSUE #2156: Make sure you want to quit
"Make sure you want to quit" message box.

### PR DESCRIPTION
**Pull request by ender-titan1**
closes #2156 
Popup doesn't show unless the last save was more than 10 seconds ago. (idk why the edits of PauseMenu.tscn are so weird, because all i did was add the timer and popup)

### CODE REVIEW

**Comment by ender-titan1** [comment]
so... i just realized i probably need to use .po files. I will convert this to a draft for now.

**Comment by FuJa0815** [comment]
> idk why the edits of PauseMenu.tscn are so weird, because all i did was add the timer and popup
thats because the indentation got changed from 4 spaces to 1 tab. Please change the indentation back (either by telling the IDE that 4 spaces should be used or using `git checkout master src/general/PauseMenu.cs` and redoing the changes). Maybe Godot messed up the indentation (See `To stop Godot from messing with you files, go to Text Editor -> Indent and set Type to spaces` section [here](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/setup_instructions.md))

**Comment by FuJa0815** [comment]
For localization, see (this)[https://github.com/Revolutionary-Games/Thrive/blob/master/doc/working_with_translations.md]

TLDR:
You need to use some kind of keyword in the scene file, something like `ARE_YOU_SURE_EXIT`
Next up you need to call the `update_localization.rb` script.
The next step is finding the new string in the `en.po` file and translating it, either by using a plain text editor or using Poedit

**Comment by ender-titan1** [comment]
So im getting a `LoadError` when trying to run `update_localization.rb`. 
`update_localization.rb:12:in 'require_relative': cannot load such file -- C:/Users/******/Thrive/RubySetupSystem/RubyCommon`, anyone knows how to fix it?

**Comment by hhyyrylainen** [comment]
You don't have submodules initialized. Looks like the update_localization script doesn't have the bootstrap call in it (if you were to run the formatting check script it automatically inits submodules). I'll make an issue for the update localization to call the bootstrap script.

**Comment by ender-titan1** [comment]
So I should just run `check_formatting.rb`, right?

**Comment by hhyyrylainen** [comment]
It should fix the issue for you...

**Comment by ender-titan1** [comment]
`ERROR: Required program / tool 'msbuild' is not installed or missing from path.
Please install it and make sure it is in path, then try again. (path is printed above for reference)`

`ERROR: Required program / tool 'inspectcode.exe' is not installed or missing from path.
Please install it and make sure it is in path, then try again. (path is printed above for reference)`

I get these when running the file. I have Visual Studio.

**Comment by hhyyrylainen** [comment]
You haven't set things up to run the code checks locally, see the setup instructions. But the script *should* have performed git submodule init before running into those issues, if it hasn't, then run git submodule init manually. Then the localize script should work.

**Comment by Kasterisk** [comment]
I'll take over this PR since there has been no updates for a while.

**Comment by Kasterisk** [comment]
Refactored code. This is now ready for review.

File: src/general/PauseMenu.cs [hunk 1]
```diff
@@ -27,11 +27,17 @@ public class PauseMenu : ControlWithInput
...
+    public NodePath ExitConfirmationDialogPath;
+
     private Control primaryMenu;
     private HelpScreen helpScreen;
     private Control loadMenu;
     private OptionsMenu optionsMenu;
     private NewSaveMenu saveMenu;
+    private ConfirmationDialog exitConfirmationDialog;
+
+    private ExitType exitType;
```
Comment (FuJa0815) [inline src/general/PauseMenu.cs:40]: add a dectiption that this is the pending exit type.

File: src/general/PauseMenu.cs [hunk 2]
```diff
@@ -191,6 +225,28 @@ private void CloseSavePressed()
...
+        switch (exitType)
+        {
+            case ExitType.ReturnToMenu:
+            {
+                // Unpause the game
+                GetTree().Paused = false;
+
+                TransitionManager.Instance.AddScreenFade(ScreenFade.FadeType.FadeOut, 0.1f, false);
+                TransitionManager.Instance.StartTransitions(this, nameof(OnSwitchToMenu));
+                break;
```
Comment (FuJa0815) [inline src/general/PauseMenu.cs:239]: Very minor, but I think extracting the cases to a seperate method helps with readability a bit.

File: locale/en.po [hunk 3]
```diff
@@ -1623,40 +1623,64 @@ msgstr "Error"
...
+msgstr "Yes"
+
+#: ../src/general/PauseMenu.cs:96
+msgid "NO"
+msgstr "No"
+
+#: ../src/general/PauseMenu.cs:160
+msgid "RETURN_TO_MENU_WARNING"
+msgstr ""
+"Are you sure you want to exit to main menu?\n"
```
Comment (FuJa0815) [inline locale/en.po:1637]:
```suggestion
"Are you sure you want to exit to the main menu?\n"
```

**Review by FuJa0815 — CHANGES_REQUESTED** [review]
looks good all in all

File: src/general/PauseMenu.tscn [hunk 4]
```diff
@@ -152,6 +155,31 @@ Category = "MicrobeStage"
 [node name="OptionsMenu" parent="." instance=ExtResource( 6 )]
 visible = false
 
+[node name="ExitConfirmation" type="ConfirmationDialog" parent="."]
+margin_right = 384.0
+margin_bottom = 84.0
+rect_min_size = Vector2( 384, 84 )
```
Comment (FuJa0815) [inline src/general/PauseMenu.tscn:161]: The popup is a bit too small for the english text

File: src/saving/SaveHelper.cs [hunk 5]
```diff
@@ -364,6 +384,8 @@ private static void PerformSave(InProgressSave inProgress, Save save)
         {
             save.SaveToFile();
             inProgress.ReportStatus(true, TranslationServer.Translate("SAVING_SUCCEEDED"));
+            SaveIsRecentlyPerformed = true;
+            recentSaveTimer.Start();
```
Comment (FuJa0815) [inline src/saving/SaveHelper.cs:388]:
```suggestion
            recentSaveTimer.Stop();
            recentSaveTimer.Start();
```
This is to reset the remaining duration if the timer is currently running

**Review by FuJa0815 — CHANGES_REQUESTED** [review]
Bug: You should reset `SaveIsRecentlyPerformed ` when loading a save/starting a new game. (`MicrobeStage.SetupStage()`)

File: src/saving/SaveHelper.cs (hunk 5 above)
Comment (Kasterisk) [reply src/saving/SaveHelper.cs:388]: I see, I was initially concerned about whether the timer automatically resets once Start() is called again after the first call (saving and then shortly saving again afterwards), but this clears it. Thanks!

File: src/general/PauseMenu.cs [hunk 6]
```diff
@@ -68,6 +87,10 @@ public override void _Ready()
         loadMenu = GetNode<Control>(LoadMenuPath);
         optionsMenu = GetNode<OptionsMenu>(OptionsMenuPath);
         saveMenu = GetNode<NewSaveMenu>(SaveMenuPath);
+        exitConfirmationDialog = GetNode<ConfirmationDialog>(ExitConfirmationDialogPath);
+
+        exitConfirmationDialog.GetOk().Text = TranslationServer.Translate("YES");
+        exitConfirmationDialog.GetCancel().Text = TranslationServer.Translate("NO");
```
Comment (FuJa0815) [inline src/general/PauseMenu.cs:93]: do this in `ShowExitConfirmationDialog`. Otherwise changing the language ingame breaks this.

**Review by FuJa0815 — CHANGES_REQUESTED** [review]

File: src/general/PauseMenu.cs (hunk 6 above)
Comment (Kasterisk) [reply src/general/PauseMenu.cs:93]: Good idea, I almost overlooked that.

**Comment by FuJa0815** [comment]
Hmm, I don't know why the CI is failing to run. It is up to date with our master...
But, local running of the scripts shows me that the localization is out of date due to the moving of code.

**Review by FuJa0815 — APPROVED** [review]

File: src/general/PauseMenu.cs (hunk 6 above)
Comment (hhyyrylainen) [reply src/general/PauseMenu.cs]: Why isn't there a separate dialog for this? That way godot automatically handles translations. Is there a reason why the confirmation dialog needs to be dynamically have the text changed?

Comment (Kasterisk) [reply src/general/PauseMenu.cs]: The ConfirmationDialog node is already suitable for this purpose but the button texts doesn't quite match given the context so it just needs a little bit tweaking to them. I guess a WindowDialog node can be used but we would have to basically readd some parts for the same functionality.

Comment (hhyyrylainen) [reply src/general/PauseMenu.cs]: I guess I'll be fine with the less duplication and problems creating alternative.

File: src/saving/SaveHelper.cs [hunk 7]
```diff
@@ -37,6 +52,12 @@ public enum SaveOrder
         FileSystem,
     }
 
+    /// <summary>
+    ///   Checks whether the last save is made within a timespan of set duration.
+    /// </summary>
+    /// <returns>True if the last save is still recent, false if otherwise or reset.</returns>
+    public static bool SaveIsRecentlyPerformed { get; private set; }
```
Comment (hhyyrylainen) [inline src/saving/SaveHelper.cs:59]: Instead of having a timer, why not store the current DateTime and compare if that's more than TimeSpan.FromSeconds(x) (with that TimeSpan being stored in Constants as a static variable)?

Comment (hhyyrylainen) [reply src/saving/SaveHelper.cs:59]: The last save needs to be nullable so the type should be `DateTime?`.

File: src/general/PauseMenu.cs [hunk 8]
```diff
@@ -237,4 +299,13 @@ private void OnSwitchToMenu()
...
         SceneManager.Instance.ReturnToMenu();
     }
+
+    private void ShowExitConfirmationDialog(string message)
+    {
+        exitConfirmationDialog.GetNode<Label>("DialogText").Text = message;
+        exitConfirmationDialog.PopupCenteredShrink();
+        exitConfirmationDialog.GetOk().ReleaseFocus();
+        exitConfirmationDialog.GetOk().Text = TranslationServer.Translate("YES");
+        exitConfirmationDialog.GetCancel().Text = TranslationServer.Translate("NO");
```
Comment (hhyyrylainen) [inline src/general/PauseMenu.cs:309]: Why are the buttons overridden? We haven't done this for other confirmation dialogs so this is inconsistent with them...

File: src/general/PauseMenu.cs [hunk 9]
```diff
@@ -49,6 +59,15 @@ public class PauseMenu : ControlWithInput
     [Signal]
     public delegate void MakeSave(string name);
 
+    /// <summary>
+    ///   Types of exit the player requests. Mainly used for game exit confirmation.
```
Comment (hhyyrylainen) [inline src/general/PauseMenu.cs:63]:
```suggestion
    ///   Types of exit the player requests. Used for game exit confirmation.
```

File: src/general/GameProperties.cs [hunk 10]
```diff
@@ -54,6 +54,8 @@ public static GameProperties StartNewMicrobeGame(bool freebuild = false)
             game.TutorialState.Enabled = false;
         }
 
+        SaveHelper.ResetSaveFreshness();
```
Comment (hhyyrylainen) [inline src/general/GameProperties.cs:57]: I don't think this should go here, This should be in the MainMenu when the scene is entered (or alternatively on both the new game and freebuild buttons).

File: locale/en.po [hunk 11]
```diff
@@ -1623,40 +1623,64 @@ msgstr "Error"
...
+#: ../src/general/PauseMenu.cs:154
+msgid "RETURN_TO_MENU_WARNING"
+msgstr ""
+"Are you sure you want to exit to the main menu?\n"
+"You will lose any unsaved progress."
+
+#: ../src/general/PauseMenu.cs:170
+msgid "QUIT_GAME_WARNING"
+msgstr ""
+"Are you sure you want to quit to the desktop?\n"
```
Comment (hhyyrylainen) [inline locale/en.po:1635]:
```suggestion
"Are you sure you want to quit the game?\n"
```
As thrive might in the future be on other platforms, this isn't accurate wording.

File: src/saving/SaveHelper.cs [hunk 12]
```diff
@@ -121,6 +142,7 @@ public static void LoadSave(string name)
     {
         GD.Print("Starting load of save: ", name);
         new InProgressLoad(name).Start();
+        ResetSaveFreshness();
```
Comment (hhyyrylainen) [inline src/saving/SaveHelper.cs:145]: I think this should do the *opposite* if you have just loaded a save and quit, the effect is basically the same as having saved just recently, so no warning about losing progress is necessary in this case.

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]

File: src/general/PauseMenu.cs (hunk 8 above)
Comment (Kasterisk) [reply src/general/PauseMenu.cs:309]: A couple of comments above: https://github.com/Revolutionary-Games/Thrive/pull/2208#discussion_r669451372.

Comment (hhyyrylainen) [reply src/general/PauseMenu.cs:309]: That comment also applies to the loading and old save etc. dialogs that also have "OK", "Cancel" buttons by default. This type of dialog would be widely usable, so maybe this should be its own class and one that reacts to locale change.
Also the translation keys should probably be changed to "YES_AGREE", NO_DISAGREE" or something as YES and NO are too general in my opinion to be usable for all languages.

Comment (FuJa0815) [reply src/general/PauseMenu.cs:309]: touched upon once #2387 is done

**Comment by hhyyrylainen** [comment]
@ender-titan1 are you planning on finishing this PR at some point?
Also if you missed my comment on one of your accepted PRs, we are currently planning a steam release, which needs different licensing. And for that to work we need previous contributors to sign our CLA: https://dev.revolutionarygamesstudio.com/cla

**Review by icedjuro — CHANGES_REQUESTED** [review]
Functionality looks good, add an English translation for "CONFIRM_EXIT"
